services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -0,0 +1,392 @@
+        }
+    }
+
+    private void unbindSecurityGroups(PortState portState) {
+
+    }
+
+    private void bindSecurityGroups(PortState portState) {
+        verifySecurityGroup(portState);
+
+        //FIXME: Not support yet
+    }
+
+    private void rollBackAllOperations(Stack<PortStateRollback> rollbacks) {
+        while (!rollbacks.isEmpty()) {
+            PortStateRollback rollback = rollbacks.pop();
+
+            try {
+                rollback.doRollback();
+            } catch (Exception e) {
+                LOG.error("{} roll back failed: {}", rollback, e);
+            }
+        }
+    }
+
+    private HostState getHostState(String hostId) {
+        return null;
+    }
+
+    private void addPortToHost(String hostId) {
+        HostState hostState = getHostState(hostId);
+
+        //FIXME: Add port to Host
+    }
+
+    public PortStateJson createPortState(String projectId, PortStateJson portStateJson) throws Exception {
+        LOG.debug("Create port state, projectId: {}, PortStateJson: {}", projectId, portStateJson);
+
+        Stack<PortStateRollback> rollbacks = new Stack<>();
+        AsyncExecutor executor = new AsyncExecutor();
+
+        PortState portState = portStateJson.getPortState();
+        portState.setProjectId(projectId);
+
+        try {
+            //Verify VPC ID
+            VpcRestWrap vpcRestWrap = new VpcRestWrap(rollbacks);
+            executor.runAsync(vpcRestWrap::verifyVpc, portState);
+
+            IpAddressRestWrap ipAddressRestWrap = new IpAddressRestWrap(rollbacks, portState.getProjectId());
+            if (portState.getFixedIps() == null) {
+                executor.runAsync(ipAddressRestWrap::allocateIpAddress, portState);
+            } else {
+                executor.runAsync(ipAddressRestWrap::verifyIpAddresses, portState.getFixedIps());
+            }
+
+            //Generate uuid for port
+            if (portState.getId() == null) {
+                portState.setId(UUID.randomUUID().toString());
+            }
+
+            MacAddressRestWrap macAddressRestWrap = new MacAddressRestWrap(rollbacks);
+            if (portState.getMacAddress() == null) {
+                executor.runAsync(macAddressRestWrap::allocateMacAddress, portState);
+            } else {
+                executor.runAsync(macAddressRestWrap::verifyMacAddress, portState);
+            }
+
+            //Verify security group
+
+            //Verify Binding Host ID
+            if (portState.getBindingHostId() != null) {
+                NodeRestWrap nodeRestWrap = new NodeRestWrap(rollbacks);
+                nodeRestWrap.verifyHost(portState.getBindingHostId());
+            }
+
+            //Wait for all async functions to finish
+            executor.joinAll();
+
+            //Persist portState
+            portRepository.addItem(portState);
+        } catch (Exception e) {
+            /**
+            When an exception occurs, we need to roll back all asynchronous operations,
```

Comment:
Hmm, this is really tricky. We will need a good Timeout story to upper bound the rollback.